src/main/scala/org/broadinstitute/dsde/rawls/expressions/ExpressionParsing.scala

```diff
@@ -145,42 +163,37 @@ class ExpressionParser extends JavaTokenParsers {
-
   private def lastAttributePipeFunc(attributeName: String)(context: ExpressionContext, graphPipeline: PipeType): FinalResult = {
-    //The below will very likely change. My suspicion is that we can't tell whether the attribute is a value
-    //or reference type without evaluating the pipe passed in and then seeing which of .getProperty or .out works.
-    //Particularly perverse graphs might have a mixture of both! (We should handle such cases cleanly, for some value of "clean".)
-    if (isValueType(attributeName)) {
-      val lastVertices = graphPipeline.toList
-
-      if (lastVertices.size() == 0) {
-        throw new RuntimeException(s"Could not dereference $attributeName because pipe returned no entities")
-      }
+    val lastVertices = graphPipeline.toList
 
-      // Look up the attributes on the list of vertices returned from the pipe. This will insert nulls into the list if
-      // the attribute doesn't exist.
-      FinalResult(lastVertices.map((v: Vertex) => {
-        v.getProperty(attributeName).asInstanceOf[Object]
-      }), s".getProperty($attributeName)")
-    } else {
-      outputEntityResult( context, graphPipeline.out(attributeName) )
+    if (lastVertices.isEmpty) {
+      throw new RuntimeException(s"Could not dereference $attributeName because pipe returned no entities")
     }
-  }
 
+    // Look up the attributes on the list of vertices returned from the pipe. This will insert nulls into the list if
+    // the attribute doesn't exist.
+    FinalResult(lastVertices.map((v: Vertex) => {
+      v.getProperty(attributeName).asInstanceOf[Object]
+    }), s".getProperty($attributeName)")
+  }
 
   //Takes a list of entities at the end of a pipeline and returns them in final format.
-  //TODO: Determine how we want to return these entities. GraphEntityDAO.loadEntity?
   private def outputEntityResult(context: ExpressionContext, graphPipeline: PipeType): FinalResult = {
-    //graphPipeline.toList.map( some_conversion_to_something_else )
-    FinalResult(List(), "")
+    val dao = new GraphEntityDAO()
+    FinalResult( graphPipeline.toList.map( dao.loadEntity(_, context.workspaceNamespace, context.workspaceName) ), "" )
   }
 }
 
 class ExpressionEvaluator(graph:Graph, parser:ExpressionParser)  {
-  def evaluate(workspaceNamespace:String, workspaceName:String, rootType:String, rootName:String, expression:String):Try[Seq[Any]] = {
-    parser.parse(expression) match {
+  def evalFinalAttribute(workspaceNamespace:String, workspaceName:String, rootType:String, rootName:String, expression:String):Try[Seq[Any]] = {
+    parser.parseAttributeExpr(expression) match {
+      case Success(pipe) => runPipe( ExpressionContext(workspaceNamespace, workspaceName, rootType, rootName), pipe)
+      case Failure(regret) => Failure(regret)
+    }
+  }
+
+  def evalFinalEntity(workspaceNamespace:String, workspaceName:String, rootType:String, rootName:String, expression:String):Try[Seq[Any]] = {
```

Comment:
yyyyyyprobably, and the attribute one could plausibly return Seq[AttributeValue] too